internal/daemon/worker/session/session.go

```diff
@@ -22,26 +22,240 @@ var errMakeSessionCloseInfoNilCloseInfo = errors.New("nil closeInfo supplied to
-	LookupSessionResponse *pbs.LookupSessionResponse
-	ConnInfoMap           map[string]*ConnInfo
+func (s *Session) GetId() string {
+	return s.resp.GetAuthorization().GetSessionId()
 }
 
-// Activate is a helper worker function that sends session activation request to the
-// controller.
-func Activate(ctx context.Context, sessClient pbs.SessionServiceClient, sessionId, tofuToken string, version uint32) (pbs.SESSIONSTATUS, error) {
+// RequestCancel sends session cancellation request to the controller.  If there is no
+// error the local session's status is updated with the result of the cancel
+// request
+func (s *Session) RequestCancel(ctx context.Context) error {
+	st, err := cancel(ctx, s.client, s.GetId())
+	if err != nil {
+		return err
+	}
+	s.ApplyLocalStatus(st)
+	return nil
+}
+
+// RequestActivate Sends session activation request to the controller.  The Session's
+// status is then updated with the result of the call.  After a successful
+// call to RequestActivate, subsequent calls will fail.
+func (s *Session) RequestActivate(ctx context.Context, tofu string) error {
+	st, err := activate(ctx, s.client, s.GetId(), tofu, s.resp.GetVersion())
+	if err != nil {
+		return err
+	}
+	s.ApplyLocalStatus(st)
+	return nil
+}
+
+// RequestAuthorizeConnection sends an AuthorizeConnection request to
+// the controller.
+// It is called by the worker handler after a connection has been received by
+// the worker, and the session has been validated.
+// The passed in context.CancelFunc is used to terminate any ongoing local proxy
+// connections.
+// The connection status is then viewable in this session's GetLocalConnections() call.
+func (s *Session) RequestAuthorizeConnection(ctx context.Context, workerId string, connCancel context.CancelFunc) (ConnInfo, int32, error) {
+	ci, connsLeft, err := authorizeConnection(ctx, s.client, workerId, s.GetId())
+	if err != nil {
+		return ConnInfo{}, connsLeft, err
+	}
+	ci.connCtxCancelFunc = connCancel
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.connInfoMap[ci.Id] = ci
+	return ConnInfo{
+		Id:        ci.Id,
+		Status:    ci.Status,
+		CloseTime: ci.CloseTime,
+	}, connsLeft, err
+}
+
+// RequestConnectConnection sends a RequestConnectConnection request to the controller. It
+// should only be called by the worker handler after a connection has been
+// authorized.  The local connection's status is updated with the result of the
+// call.
+func (s *Session) RequestConnectConnection(ctx context.Context, info *pbs.ConnectConnectionRequest) error {
+	st, err := connectConnection(ctx, s.client, info)
+	if err != nil {
+		return err
+	}
+	s.ApplyLocalConnectionStatus(info.GetConnectionId(), st)
+	return nil
+}
+
+// CancelOpenLocalConnections closes the local connections in this session
+//based on the connection's state by calling the connections context cancel
+// function.
+//
+// The returned slice are connection ids that were closed.
+func (s *Session) CancelOpenLocalConnections() []string {
```

Comment:
I'm not sure this and the function below need to be different. Seems like CancelOpenLocalConnections should suffice for any need?